### PR TITLE
8298642: ParallelGC -XX:+UseNUMA eden spaces allocated on wrong node

### PR DESCRIPTION
**Pull request by nick-arm**
I noticed this when implementing JDK-8298482.  If UseAdaptiveSizePolicy and/or UseAdaptiveNUMAChunkSizing are enabled and UseLargePages is *not* enabled, part of the eden spaces managed by MutableNUMASpace end up bound to the wrong node.  This seems to be a regression caused by JDK-8283935, specifically the addition of

```
  set_page_size(alignment());
```

to MutableNUMASpace::initialize().  Previously the page size would have been equal to os::vm_page_size() unless UseLargePages was enabled.

In bias_region() we call os::free_memory() to uncommit a range of pages so that they can be later allocated on a different node, passing page_size() as the alignment_hint argument. However on Linux this does nothing if alignment_hint is greater than os::vm_page_size() and THP is not enabled, so the pages always remain bound to their original node (see os::pd_free_memory() in os_linux.cpp).

I think the solution here is to pass os::vm_page_size() as the alignment_hint when UseLargePages is disabled.

I tested SPECjbb on a dual-socket AArch64 system with -XX:+UseParallelGC -XX:+UseNUMA but otherwise default options and critical-jOPS improved by about 12% (averaged over several runs).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298642](https://bugs.openjdk.org/browse/JDK-8298642): ParallelGC -XX:+UseNUMA eden spaces allocated on wrong node

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk20 pull/33/head:pull/33` \
`$ git checkout pull/33`

Update a local copy of the PR: \
`$ git checkout pull/33` \
`$ git pull https://git.openjdk.org/jdk20 pull/33/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 33`

View PR using the GUI difftool: \
`$ git pr show -t 33`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk20/pull/33.diff">https://git.openjdk.org/jdk20/pull/33.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — APPROVED** [review]
I think this is good. Please wait for a second review.

**Review by albertnetymk — APPROVED** [review]

**Comment by albertnetymk** [comment]
> In bias_region() we call os::free_memory() to uncommit a range of pages

Just for the record, `os::free_memory()` does not uncommit any mem. Instead, it calls `commit_memory` to overwrite existing mmapings.

I believe the reason why this fix works is that at startup eden-space is committed (within young-gen) using a single `mmap` call (`PSVirtualSpace::expand_by`), resulting into one mmapping, which is later overwritten (in `MutableNUMASpace::bias_region`) to 2 (assuming 2 numa nodes) mmappings that can bind to diff numa nodes.

**Comment by nick-arm** [comment]
> > In bias_region() we call os::free_memory() to uncommit a range of pages
> 
> Just for the record, `os::free_memory()` does not uncommit any mem. Instead, it calls `commit_memory` to overwrite existing mmapings.

Doesn't it? I think `mmap`-ing over an existing mapping has the effect of freeing the physical pages backing it, allowing them to be allocated again on a different node the next time they are touched. Otherwise setting the NUMA node hint with `os::numa_make_local` doesn't have any effect if physical pages are already allocated as Linux won't migrate pages unless you call `move_pages`. Certainly if I call `os::numa_get_group_id_for_address` after `os::free_memory` it always returns -1 (not allocated on any node).

**Comment by albertnetymk** [comment]
> Doesn't it? I think mmap-ing over an existing mapping has the effect of freeing the physical pages backing it,

Yes, you are right. I tested just now using `mincore` that overwriting existing mmapings releases the underlying physical mem.

**Comment by nick-arm** [comment]
Thanks for the reviews!

**Pull request closed without merging**
